runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java

```diff
@@ -0,0 +1,255 @@
+     */
+    boolean canFuse(PTransformNode consumer, ExecutableStage subgraph, QueryablePipeline pipeline);
+  }
+
+  private interface CompatibilityChecker {
+    /**
+     * Determine if two {@link PTransformNode PTransforms} can be fused into a new stage. This
+     * determines sibling fusion for new {@link ExecutableStage stages}.
+     */
+    boolean isCompatible(
+        PTransformNode newNode, PTransformNode otherNode, QueryablePipeline pipeline);
+  }
+
+  /**
+   * A ParDo can be fused into a stage if it executes in the same Environment as that stage, and no
+   * transform that are upstream of any of its side input are present in that stage.
+   *
+   * <p>A ParDo that consumes a side input cannot process an element until all of the side inputs
+   * contain data for the side input window that contains the element.
+   */
+  private static boolean canFuseParDo(
+      PTransformNode parDo, ExecutableStage stage, QueryablePipeline pipeline) {
+    Optional<Environment> env = pipeline.getEnvironment(parDo);
+    checkArgument(
+        env.isPresent(),
+        "A %s must have an %s associated with it",
+        ParDoPayload.class.getSimpleName(),
+        Environment.class.getSimpleName());
+    if (!env.get().equals(stage.getEnvironment())) {
+      // The PCollection's producer and this ParDo execute in different environments, so fusion
+      // is never possible.
+      return false;
+    } else if (!pipeline.getSideInputs(parDo).isEmpty()) {
+      // At execution time, a Runner is required to only provide inputs to a PTransform that, at the
+      // time the PTransform processes them, the associated window is ready in all side inputs that
+      // the PTransform consumes. For an arbitrary stage, it is significantly complex for the runner
+      // to determine this for each input. As a result, we break fusion to simplify this inspection.
+      // In general, a ParDo which consumes side inputs cannot be fused into an executable subgraph
+      // alongside any transforms which are upstream of any of its side inputs.
+      return false;
+    }
+    return true;
+  }
+
+  private static boolean parDoCompatibility(
+      PTransformNode parDo, PTransformNode other, QueryablePipeline pipeline) {
+    if (!pipeline.getSideInputs(parDo).isEmpty()) {
+      // This is a convenience rather than a strict requirement. In general, a ParDo that consumes
+      // side inputs can be fused with other transforms in the same environment which are not
+      // upstream of any of the side inputs.
+      return false;
+    }
+    return compatibleEnvironments(parDo, other, pipeline);
+  }
+
+  /**
+   * A WindowInto can be fused into a stage if it executes in the same Environment as that stage.
+   */
+  private static boolean canFuseWindow(
+      PTransformNode window, ExecutableStage stage, QueryablePipeline pipeline) {
+    // WindowInto transforms may not have an environment
```

Comment:
When would this happen?

Reply:
For a pipeline to be portable, it can't assume that a Runner will override something otherwise that SDK won't be usable on other runners who don't do that as well.

Reply:
For a pipeline to be portable, it can't assume that a runner will perform an override otherwise other runners will need to follow suit which is why an environment is required.